osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/CatchUpSpectatorPlayerClock.cs

```diff
@@ -66,7 +66,7 @@ public void ProcessFrame()
 
             if (IsRunning)
             {
-                double elapsedSource = Source.ElapsedFrameTime;
+                double elapsedSource = Math.Max(Source.ElapsedFrameTime, 0);
```

Comment:
Curious why this isn't an issue when rewinding..

Reply:
Yeah, now that I re-read this again and the reason why, this particular code is the cause of this:
https://github.com/ppy/osu/blob/543482111b470d63cf74c4ed6dab294282fa7af8/osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs#L185-L186

When a clock offset is set, the initial seek on master to the spectator first frame time will instead make `CurrentTime` end up becoming `time + offset` rather than `time`, and then when the sync manager realizes that master is ahead of spectator, it stops master which leads to `CurrentTime` slowly decreasing back from `CurrentTime + offset` to `CurrentTime` (due to `Rate` influencing the offset clock), making the sync manager think it's catched up already and starts it again, and the loop continues.

This is actually a regression from [this change](https://github.com/ppy/osu/pull/14253/commits/b1e3327f979dd779bdc3be36e2177317a5f6ed8c) which corrected the seek on different rates (basically nullifying what `HardwareCorrectionOffsetClock` does for the current time to match the proposed time).

Maybe `MasterGameplayClockContainer.Seek` should ignore the `Rate` if the master clock is not running? seems like a workaround though...

Flipping the multi-spectator screen code to first start then seek also won't work, because `MasterGameplayClockContainer` applies a 200ms transform to the rate on pause change, perhaps it's incorrect for that to happen in the first place?

Reply:
I'm not sure I follow your logic here. How does offset and rate relate to each other..?

Reply:
Because of this:
https://github.com/ppy/osu/blob/009aa994d09eca2508af2f3d4ec4fb73fae3a7f6/osu.Game/Screens/Play/MasterGameplayClockContainer.cs#L215-L225

Reply:
This adjustment should *only* be applied to the master clock. It's intended to adjust the timing of the audio track sync, and doesn't need to be applied at each spectator client. Is that not what's currently happening?

Reply:
Yes it's only applied to the master clock, but the fact that it changes between `time` and `time + Offset` based on the clock state becomes a problem, like in the scenario I mentioned.
> https://github.com/ppy/osu/blob/543482111b470d63cf74c4ed6dab294282fa7af8/osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs#L185-L186
> 
> When a clock offset is set, the initial seek on master to the spectator first frame time will instead make `CurrentTime` end up becoming `time + offset` rather than `time`, and then when the sync manager realizes that master is ahead of spectator, it stops master which leads to `CurrentTime` decreasing back from `time + offset` to `time` (due to `Rate` influencing the offset clock), making the sync manager think it's catched up already and starts it again, and the loop continues.

Reply:
Actually, now that I wrote all of this, I've came to a solution that renders this current one highlighted by the conversation as completely invalid.

To explain this once more, you first have to consider that `HardwareCorrectionOffsetClock` manipulates the `CurrentTime` based on the `Rate`, so when a clock stops and a pause frequency adjustment is applied, the `Rate` slowly changes from 1 to 0, manipulating the current time of the clock and causing issues to the sync manager.

So in the above scenario I mentioned, when the master clock has a user offset applied (say 80ms in the following example), the following happens on a regular multi-spectate session:
 - The master clock container gets stopped and reset.
 - The multi-spectator player receives the first frame bundle, and calls `Seek` on the provided frame time.
 - The sync manager acknowledges that and fires up `ReadyToStart` action, which `MultiSpectatorScreen` handles and proceeds to seek the master clock and starts it up.
 - The master clock ends up seeking to 80ms instead, due to the `Rate` being 0 when the seek was performed.
      - This was not the case before https://github.com/ppy/osu/pull/14253/commits/b1e3327f979dd779bdc3be36e2177317a5f6ed8c, but I've pushed that change as that code was resulting in multiple incorrect seeks when `Start()` or `Seek()` gets called while the clock is stopped, making the master clock go behind or ahead of the provided time.
 - The sync manager notices that and stops the master clock for the spectator clock to catch up (after few iteration of frames, in which `GameplayClock.UnderlyingClock.LastFrameTime` would've already updated to 80ms as the frequency adjustment transformation goes).
 - While the master clock stops, the `Rate` interpolates back from one to zero, manipulating the `CurrentTime` to become 0ms, `GameplayClock.ElapsedFrameTime` becomes a negative value and since the master clock is stopped, the elapsed frame time never updates until it's started again.
 - The spectator clock reads that negative `ElapsedFrameTime` and advances itself by it thinking it's catching up, while it's rewinding away forever.